File: lib-androscope/src/main/java/nl/ngti/androscope/server/ResponseFactory.java
package nl.ngti.androscope.server;

import android.content.Context;
import android.os.Bundle;
import android.util.Log;

import androidx.annotation.Nullable;

import java.io.IOException;

import fi.iki.elonen.NanoHTTPD;
import nl.ngti.androscope.common.AndroscopeConstants;

final class ResponseFactory {

    private static final boolean LOG = AndroscopeConstants.LOG;
    private static final String TAG = ResponseFactory.class.getSimpleName();

    private final NotFoundResponse mNotFoundResponse = new NotFoundResponse();
    private final AssetResponse mAssetResponse = new AssetResponse();
    private final RestResponse mRestResponse = new RestResponse();

    ResponseFactory(Context context, Bundle metadata) {
        mNotFoundResponse.init(context, metadata);
        mAssetResponse.init(context, metadata);
        mRestResponse.init(context, metadata);
    }

    @Nullable
    NanoHTTPD.Response getResponse(NanoHTTPD.IHTTPSession session) throws IOException {
        final String uri = session.getUri();
        if (LOG) Log.d(TAG, "getResponse " + uri + ", params " + session.getParameters());

        if (uri.startsWith(AndroscopeConstants.PATH_REST)) {
            return mRestResponse.getResponse(new SessionParams(session, AndroscopeConstants.PATH_REST));
        }

        if (uri.startsWith("/")) {
            return mAssetResponse.getResponse(new SessionParams(session, "/"));
        }

        if (LOG) Log.w(TAG, "No response to handle " + uri);
        return mNotFoundResponse.getResponse(new SessionParams(session, "/"));
    }
}

File: lib-androscope/src/main/java/nl/ngti/androscope/server/ResponseFactory.kt
package nl.ngti.androscope.server

import android.content.ContentResolver
import android.content.Context
import android.net.Uri
import fi.iki.elonen.NanoHTTPD
import fi.iki.elonen.NanoHTTPD.IHTTPSession
import nl.ngti.androscope.common.log
import nl.ngti.androscope.common.logTime
import nl.ngti.androscope.responses.common.MultiSchemeDataProvider
import nl.ngti.androscope.responses.common.UriDataProvider
import nl.ngti.androscope.utils.AndroscopeMetadata

internal class ResponseFactory(
        context: Context,
        metadata: AndroscopeMetadata
) {

    private val uriDataProvider = MultiSchemeDataProvider().apply {
        addProvider(ContentResolver.SCHEME_CONTENT, ContentResolverDataProvider(context.contentResolver))
    }

    private val urlMatcher = ResponseMatcher(context, metadata, uriDataProvider)

    fun getResponse(session: IHTTPSession): NanoHTTPD.Response? {
        log {
            """getResponse
                | path ${session.path},
                | params ${session.parameters}""".trimMargin()
        }
        val handler = urlMatcher[session]

        return logTime({
            handler(session)
        }, {
            "Response generation time [${session.path}]"
        })
    }
}

private class ContentResolverDataProvider(
        private val contentResolver: ContentResolver
) : UriDataProvider {

    override fun query(
            uri: Uri,
            projection: Array<String>?,
            selection: String?,
            selectionArgs: Array<String>?,
            sortOrder: String?
    ) = contentResolver.query(uri, projection, selection, selectionArgs, sortOrder)
}
